app/uk/gov/hmrc/essttpstubs/controllers/PegaController.scala

```diff
@@ -16,37 +16,89 @@
 
+import java.time.{LocalDateTime, ZoneOffset}
 import javax.inject.{Inject, Singleton}
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.Random
 
 @Singleton
-class PegaController @Inject() (cc: ControllerComponents) extends BackendController(cc) {
+class PegaController @Inject() (cc: ControllerComponents, repo: PegaTokenRepo, appConfig: AppConfig)(implicit ec: ExecutionContext) extends BackendController(cc) {
 
   val logger: Logger = Logger(this.getClass)
 
-  val token: Action[AnyContent] = Action { implicit request =>
-    val response = Json.parse(
+  private val pegaTokenExpiryTime: FiniteDuration = appConfig.pegaTokenExpiryTime
+
+  val token: Action[AnyContent] = Action.async { implicit request =>
+
+    repo.findPegaToken().flatMap {
+      case None =>
+        generateAndSaveToken()
+
+      case Some(pegaToken) =>
+        val expirationTime = pegaToken.validFrom.plusSeconds(pegaTokenExpiryTime.toSeconds)
+        if (LocalDateTime.now().isAfter(expirationTime)) {
+          generateAndSaveToken()
+        } else {
+          val remainingTime = expirationTime.toEpochSecond(ZoneOffset.UTC) - LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)
+          val response = generateTokenResponse(pegaToken.accessToken, remainingTime.toInt)
+          logRequest(request, response)
+          Future.successful(Ok(response))
+        }
+    }
+  }
+
+  private def generateAndSaveToken()(implicit request: Request[AnyContent]): Future[Result] = {
+    val newToken = PegaOauthToken(nextAlphanumericString(20), LocalDateTime.now())
+    repo.insertPegaToken(newToken).map { _ =>
+      val response = generateTokenResponse(newToken.accessToken, pegaTokenExpiryTime.toSeconds * 2)
+      logRequest(request, response)
+      Ok(response)
+    }
+  }
+
+  private def generateTokenResponse(accessToken: String, expiresIn: Long): JsValue = {
+    Json.parse(
       s"""{
-        |  "access_token": "${nextAlphanumericString(20)}",
-        |  "token_type": "bearer",
-        |  "expires_in": 360000
-        |}
-        |""".stripMargin
+         |  "access_token": "$accessToken",
+         |  "token_type": "bearer",
+         |  "expires_in": ${expiresIn.toString}
+         |}
+         |""".stripMargin
     )
+  }
+
+  private def logRequest(request: Request[AnyContent], response: JsValue): Unit = {
     logger.info(
       s"Got request for PEGA token with headers ${request.headers.headers.toString} and " +
         s"body ${request.body.toString}. Responding with ${response.toString}"
     )
-    Ok(response)
   }
 
-  val startCase: Action[AnyContent] = Action { implicit request =>
-    val response = Json.parse(
+  val startCase: Action[AnyContent] = Action.async { implicit request =>
+    val maybeAuthToken = request.headers.get("Authorization").flatMap(parseBearerToken)
+
+    validateToken(maybeAuthToken).flatMap {
```

Comment:
you could probably just pass in the request to `validateToken` and stick
```
val maybeAuthToken = request.headers.get("Authorization").flatMap(parseBearerToken
``` 
in there instead of repeating it here and in `getCase`